### PR TITLE
(3/5) Multiple Issues: Close gaps after mouse button release

### PR DESCRIPTION
**Pull request by andreasfelix**
**(3/5) Split up #177 in different PRs:**

Needs to be merged after #181

**Changes in this PR:**
- close gaps after mouse button release for up to 3 monitors (#35.1)

close-gaps

### CODE REVIEW

File: src/Widgets/DisplaysOverlay.vala [hunk 1]
```diff
@@ -275,6 +281,47 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
...
+                continue;
+            }
+
+            int other_x, other_y, other_width, other_height;
+            other_display_widget.get_geometry (out other_x, out other_y, out other_width, out other_height);
+
+            Gdk.Rectangle other_rect = { other_x, other_y, other_width, other_height };
+            Gdk.Rectangle intersection;
+            var is_connected = rect.intersect (other_rect, out intersection);
+            var is_diagonal = intersection.height == 1 && intersection.width == 1;
```
Comment (jeremypw) [inline src/Widgets/DisplaysOverlay.vala]: Should be `var is_diagonal = is_connected && (intersection.height <= 50 && intersection.width <= 50);` to regard "forbidden" formation as disconnected. Probably best to make the `50` a const variable for clarity.

File: src/Widgets/DisplaysOverlay.vala [hunk 2]
```diff
@@ -384,116 +431,62 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
...
+            if (distance_left > 0 && distance_right < 0) {
+                test_distance_x = 0;
+            } else if (distance_top > 0 && distance_bottom < 0) {
+                test_distance_y = 0;
+            } else { // As diagonal monitors are not allowed, offset by 50px
+                if (test_distance_x.abs () >= test_distance_y.abs ()) {
+                    test_distance_x += diff_x > 0 ? 50 : -50;
+                } else {
+                    test_distance_y += diff_y > 0 ? 50 : -50;
+                }
```
[the remark this answers is not in the document]
Comment (andreasfelix) [reply src/Widgets/DisplaysOverlay.vala]: I renamed diff_x/y to distance_widget_anchor_x/y to make variable names more explanatory. Do you have some other suggestions for better variable names?

I think the difficulty to name geometry stuff is that there are 100 different equivalent ways to look at same problem. So a formula which seem intuitive someone might appear counter-intuitive to someone else...

**Comment by andreasfelix** [comment]
@jeremypw I intentional used the 50px offset only in the snapping function. My idea was that it seems like a reasonable default and makes it clear to the user that diagonal monitors are not allowed (1px would be hard to notice). But if a user wants to place their monitor within the 1-49 px margin (which technically allowed), I see no reason to explicitly forbid this.

But if you have a strong opinion that the 50px offset should be mandatory, I will change it. I just wanted to let you know that I already took it into consideration...

File: src/Widgets/DisplaysOverlay.vala (hunk 1 above)
Comment (andreasfelix) [reply src/Widgets/DisplaysOverlay.vala]: - made 50 a const
- for first point see comment

**Comment by jeremypw** [comment]
@andreasfelix Regarding diagonal offsets, I am happy to go with your way for now.  No doubt an issue will be raised if there is any serious objection. 
With regards to comments, all I can say is that I found it hard to understand the code without adding comments of my own to remind me what each variable represented.  Variables `distance_x` and `test_distance_x` etc are hard to understand without a comment. A picture would be even better! Variable names of reasonable length that are fully self documenting does not come to mind I am afraid.

File: src/Widgets/DisplaysOverlay.vala [hunk 3]
```diff
@@ -384,116 +433,62 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
...
+            var distance_widget_anchor_x = anchor_x - widget_x;
+            var distance_widget_anchor_y = anchor_y - widget_y;
+            var distance_left   = distance_widget_anchor_x + anchor_width;
+            var distance_right  = distance_widget_anchor_x - widget_width;
+            var distance_top    = distance_widget_anchor_y + anchor_height;
+            var distance_bottom = distance_widget_anchor_y - widget_height;
+            var test_distance_x = distance_right > -distance_left ? distance_right : distance_left;
+            var test_distance_y = distance_bottom > -distance_top ? distance_bottom : distance_top;
+
+            if (distance_left > 0 && distance_right < 0) {
```
Comment (jeremypw) [inline src/Widgets/DisplaysOverlay.vala]: Comment that this meams widget is between left and right edges of anchor so no horizontal movement needed.

File: src/Widgets/DisplaysOverlay.vala [hunk 4]
```diff
@@ -384,116 +433,62 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
...
+                test_distance_x = 0;
+            } else if (distance_top > 0 && distance_bottom < 0) {
```
Comment (jeremypw) [inline src/Widgets/DisplaysOverlay.vala]: Comment that this meams widget is between top and bottom edges of anchor so no vertical movement needed.

File: src/Widgets/DisplaysOverlay.vala (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremypw) [reply src/Widgets/DisplaysOverlay.vala]: Then comments to indicate the intention of each formula are needed.

File: src/Widgets/DisplaysOverlay.vala [hunk 5]
```diff
@@ -384,116 +433,62 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
...
-            if (snap_y & move) {
-                debug ("moving child %d on Y\n", shortest_y);
-                if (shortest_y < SNAP_LIMIT) ((DisplayWidget) child).set_geometry (child_x , child_y - shortest_y, child_width, child_height);
+            var distance_widget_anchor_x = anchor_x - widget_x;
+            var distance_widget_anchor_y = anchor_y - widget_y;
+            var distance_left   = distance_widget_anchor_x + anchor_width;
+            var distance_right  = distance_widget_anchor_x - widget_width;
+            var distance_top    = distance_widget_anchor_y + anchor_height;
+            var distance_bottom = distance_widget_anchor_y - widget_height;
+            var test_distance_x = distance_right > -distance_left ? distance_right : distance_left;
```
Comment (jeremypw) [inline src/Widgets/DisplaysOverlay.vala]: Isn't this the same as 
`var test_distance_x = distance_right.abs () < distance_left.abs () ? distance_right : distance_left;`

If so, it is more obvious that `test_distance_x` is in fact the smallest horizontal gap and could perhaps be renamed as such.

Comment (andreasfelix) [reply src/Widgets/DisplaysOverlay.vala]: - Yes, it is the same
- I renamed `test_distance_x`

**Comment by andreasfelix** [comment]
@jeremypw, I have added the comments where it is checked if the widget is within the edges of the anchor. I have also renamed the `distance_x/y` and `test_distance_x/y` variables. I think their purpose should be more clear now.

If there are any other parts which are difficult to understand, I think it would be the best (if it is okay for you) if you would comment them. For me it is not obvious what could be unclear and I am afraid I would add to many unnecessary comments.

File: src/Widgets/DisplaysOverlay.vala [hunk 6]
```diff
@@ -275,6 +282,48 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
         check_intersects (display_widget);
     }
 
+    private void close_gaps () {
+        var display_widgets = new List<DisplayWidget>();
```
Comment (jeremypw) [inline src/Widgets/DisplaysOverlay.vala]: Lint: missing space

File: src/Widgets/DisplaysOverlay.vala [hunk 7]
```diff
@@ -384,116 +433,65 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
...
-                case_3 = is_x_smaller_absolute (case_3, case_3_t) && !diagonally ? case_3 : case_3_t;
-                case_4 = is_x_smaller_absolute (case_4, case_4_t) && !diagonally ? case_4 : case_4_t;
-                snap_y = true;
-                move = true;
+
+            var distance_origin_x = anchor_x - widget_x;
+            var distance_origin_y = anchor_y - widget_y;
+            var distance_left   = distance_origin_x + anchor_width;
+            var distance_right  = distance_origin_x - widget_width;
+            var distance_top    = distance_origin_y + anchor_height;
```
Comment (jeremypw) [inline src/Widgets/DisplaysOverlay.vala]: Lint: multiple spaces instead of single space in above 3 lines.

File: src/Widgets/DisplaysOverlay.vala [hunk 8]
```diff
@@ -384,116 +433,65 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
...
 
-        int shortest_x = is_x_smaller_absolute (case_1, case_2) ? case_1 : case_2;
-        int shortest_y = is_x_smaller_absolute (case_3, case_4) ? case_3 : case_4;
-
-        // X Snapping
-        if (!snap_y || is_x_smaller_absolute (shortest_x, shortest_y)) {
-            if (snap_x & move) {
-                debug ("moving child %d on X\n", shortest_x);
-                if (shortest_x < SNAP_LIMIT) ((DisplayWidget) child).set_geometry (child_x - shortest_x, child_y , child_width, child_height);
+            var shortest_distance_candidate = distance_widget_anchor_x * distance_widget_anchor_x + distance_widget_anchor_y * distance_widget_anchor_y;
```
Comment (jeremypw) [inline src/Widgets/DisplaysOverlay.vala]: Lint: Significantly exceeds 120 char line length

**Pull request merged**